jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IteratingCallback.java

```diff
@@ -198,6 +213,93 @@ protected void onCompleteFailure(Throwable cause)
     {
     }
 
+    /**
+     * Invoked when the overall task has been aborted.
+     * <p>
+     * Calls to this method are serialized with respect to {@link #process()}, {@link #onCompleteFailure(Throwable)}
+     * and {@link #onCompleted(Throwable)}.
+     * If this method is called, then {@link #onCompleteSuccess()} will never be called.
+     * <p>
+     * The default implementation of this method calls {@link #failed(Throwable)}.  Overridden implementations of
+     * this method SHOULD NOT call {@code super.onAborted(Throwable)}.
+     *
+     * @param cause The cause of the abort
+     */
+    protected void onAborted(Throwable cause)
+    {
+    }
+
+    /**
+     * Invoked when the overall task has completed.
+     * <p>
+     * Calls to this method are serialized with respect to {@link #process()} and {@link #onAborted(Throwable)}.
+     * The default implementation of this method will call either {@link #onCompleteSuccess()} or {@link #onCompleteFailure(Throwable)}
+     * thus implementations of this method should always call {@code super.onCompleted(Throwable)}.
+     *
+     * @param causeOrNull the cause of any {@link #abort(Throwable) abort} or {@link #failed(Throwable) failure},
+     * else {@code null} for {@link #succeeded() success}.
+     */
+    protected void onCompleted(Throwable causeOrNull)
+    {
+        if (causeOrNull == null)
+            onCompleteSuccess();
+        else
+            onCompleteFailure(causeOrNull);
+    }
+
+    private void doOnSuccessProcessing()
+    {
+        ExceptionUtil.callAndThen(this::onSuccess, this::processing);
```

Comment:
@lorban @sbordet Should I inline this call to avoid capturing `this` twice?   This is the hot path, the calls below are all error paths, but maybe we should inline them all?

Reply:
Yes make `this::onSuccess` a field so it's captured once.